internal/gather/gather_bytes_test.go

```diff
@@ -173,6 +182,119 @@ func TestGatherBytesReadSeeker(t *testing.T) {
 	require.Error(t, err)
 }
 
+func TestGatherBytesReaderAtErrorResponses(t *testing.T) {
+	// 3.7 times the internal chunk size
+	contentBuf := make([]byte, int(float64(defaultAllocator.chunkSize)*3.7))
+	for i := range contentBuf {
+		contentBuf[i] = uint8(i % math.MaxUint8)
+	}
+
+	tcs := []struct {
+		inBsLen   int
+		inOff     int64
+		expectErr error
+		expectN   int
+	}{
+		{
+			inBsLen:   1 << 10,
+			inOff:     -1,
+			expectErr: ErrInvalidOffset,
+			expectN:   0,
+		},
+		{
+			inBsLen:   1 << 10,
+			inOff:     math.MaxInt64,
+			expectErr: io.EOF,
+			expectN:   0,
+		},
+		{
+			inBsLen:   0,
+			inOff:     -1,
+			expectErr: ErrInvalidOffset,
+			expectN:   0,
+		},
+		{
+			inBsLen: 0,
+			inOff:   math.MaxInt64,
+			expectN: 0,
+		},
+	}
+	for i, tc := range tcs {
+		t.Run(fmt.Sprintf("%d: %d %d %d", i, tc.inBsLen, tc.inOff, tc.expectN), func(t *testing.T) {
+			// tmp is an empty buffer that will supply some bytes
+			// for testing
+			var wrt WriteBuffer
+			defer wrt.Close()
+
+			wrt.Append(contentBuf)
+			require.Equalf(t, defaultAllocator.chunkSize, wrt.alloc.chunkSize,
+				"this test expects that the default-allocator will be used, but we are using: %#v", wrt.alloc)
+
+			// get the reader out of the WriteBuffer so we can read what was written
+			// (presume all 0s)
+			reader := wrt.inner.Reader()
+			defer reader.Close() //nolint:errcheck
+
+			// get the reader as a ReaderAt
+			readerAt := reader.(io.ReaderAt)
+
+			// make an output buffer of the required length
+			bs := make([]byte, tc.inBsLen)
+
+			n, err := readerAt.ReadAt(bs, tc.inOff)
+			require.ErrorIs(t, err, tc.expectErr)
+			require.Equal(t, tc.expectN, n)
+		})
+	}
+}
+
+func TestGatherBytesReaderAtVariableInputBufferSizes(t *testing.T) {
+	const inputBufferMaxMultiplier = 4.0 // maximum number of times the internal chunk size
+
+	contentBuf := make([]byte, defaultAllocator.chunkSize*inputBufferMaxMultiplier)
+	for i := range contentBuf {
+		contentBuf[i] = uint8(i % math.MaxUint8)
+	}
+
+	// Test some interesting input buffer sizes from a minimum three
+	// times the size of the internal allocator byte slices.
+	//
+	// The idea here is to exercise the part of the buffer ReaderAt
+	// implementation where it has a longer buffer size than the size of the
+	// internal chunks of the buffer implementation. When we do this, the
+	// ReaderAt is forced to draw more data than it actually can from the
+	// first slice it found after searching for the current pointer in read
+	// cycle. Finally, it should increment the read index correctly.
+	//
+	// x.1 ... x.9
```

Comment:
So IIUC, we're testing contentBufs with size `[3.1x, 3.2x, ..., 3.9x]` (x is the chunk size)? I was thinking we could try more interesting contentBuf sizes like `[0.5x, 1x, 2x, 2x-1, 2x+1,...]` that sort of thing (x is the chunk size). That would probe around the edges of the different regimes involved in the underlying buffer implementation. The current test just exercises the "multiple slices with a partial slice at the end" case.

Reply:
done